pkg/controller/actions/gc/action_gc.go

```diff
@@ -0,0 +1,126 @@
+package gc
+
+import (
+	"context"
+	"fmt"
+	"slices"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	"github.com/opendatahub-io/opendatahub-operator/v2/pkg/controller/actions"
+	odhTypes "github.com/opendatahub-io/opendatahub-operator/v2/pkg/controller/types"
+	"github.com/opendatahub-io/opendatahub-operator/v2/pkg/services/gc"
+)
+
+type PredicateFn func(*odhTypes.ReconciliationRequest, unstructured.Unstructured) (bool, error)
+type ActionOpts func(*Action)
+
+type Action struct {
+	labels       map[string]string
+	selector     labels.Selector
+	unremovables []schema.GroupVersionKind
+	gc           *gc.GC
+	predicateFn  PredicateFn
+}
+
+func WithLabel(name string, value string) ActionOpts {
+	return func(action *Action) {
+		if action.labels == nil {
+			action.labels = map[string]string{}
+		}
+
+		action.labels[name] = value
+	}
+}
+
+func WithLabels(values map[string]string) ActionOpts {
+	return func(action *Action) {
+		if action.labels == nil {
+			action.labels = map[string]string{}
+		}
+
+		for k, v := range values {
+			action.labels[k] = v
+		}
+	}
+}
+
+func WithUnremovables(items ...schema.GroupVersionKind) ActionOpts {
+	return func(action *Action) {
+		action.unremovables = append(action.unremovables, items...)
+	}
+}
+
+func WithPredicate(value PredicateFn) ActionOpts {
+	return func(action *Action) {
+		if value == nil {
+			return
+		}
+
+		action.predicateFn = value
+	}
+}
+
+func WithGC(value *gc.GC) ActionOpts {
+	return func(action *Action) {
+		if value == nil {
+			return
+		}
+
+		action.gc = value
+	}
+}
+
+func (a *Action) run(ctx context.Context, rr *odhTypes.ReconciliationRequest) error {
+	// To avoid the expensive GC, run it only when resources have
+	// been generated
+	if !rr.Generated {
+		return nil
+	}
```

Comment:
so this is saying:
even component CR gets deleted by human, after it gets recreated, since we have cache, `rr.Generated = true` wont be called, thus this run() will get fast return nil

Reply:
Not sure if I got it eight, but in the case of the component CR being re-created, the `rr.Generated == true` because the rendering task takes into account also the CR's UID when computing the hash so the gc task would kicks in.